src/caffeine/caffeine.c

```diff
@@ -179,15 +201,97 @@ void caf_get_strided(int dims, int image_num,
 
 //-------------------------------------------------------------------
 
+// caf_segment_release() is invoked whenever this image is ending a
+// segment, to flush any pending actions that are specified to be
+// ordered before a subsequent segment.
+void caf_segment_release() {
+  // synchronize caf_event_post:
+  gex_NBI_Wait(GEX_EC_RMW, 0);
+}
+
 void caf_sync_memory() {
-  // we may eventually need more than this if/when we relax our memory model..
+  caf_segment_release();
+
   gasnett_local_mb();
 }
 
 void caf_sync_team( gex_TM_t team ) {
+  caf_segment_release();
+
   gex_Event_Wait( gex_Coll_BarrierNB(team, 0) );
 }
 
+// _______________________ Events ____________________________
+
+static gex_AD_t event_AD = GEX_AD_INVALID;
+
+static void event_init(void) {
+  assert(event_AD == GEX_AD_INVALID);
+
+  // create the event AD and request CPU/AM transport
+  gex_AD_Create(&event_AD, myworldteam, GEX_DT_I64, 
+                GEX_OP_GET | GEX_OP_INC | GEX_OP_FSUB, 
+                GEX_FLAG_AD_FAVOR_MY_RANK);
+
+  assert(event_AD != GEX_AD_INVALID);
+}
+
+void caf_event_post(int image, intptr_t event_var_ptr, int segment_boundary, int release_fence) {
+  assert(event_AD != GEX_AD_INVALID);
+  assert(event_var_ptr);
+
+  if (segment_boundary) {
+    caf_segment_release();
+  }
+
+  gex_AD_OpNBI_I64(event_AD, NULL, 
+                   image-1, (void *)event_var_ptr, 
+                   GEX_OP_INC, 0, 0, 
+                   (release_fence ? GEX_FLAG_AD_REL : 0));
+
+  // We've issued the post increment as an NBI operation,
+  // allowing this call to return before the increment
+  // is acknowledged by the remote side.
+  // This will later be synchronized in caf_segment_release()
+}
+
+void caf_event_query(void *event_var_ptr, int64_t *count) {
+  assert(event_AD != GEX_AD_INVALID);
+  assert(event_var_ptr);
+  assert(count);
+
+  gex_Event_Wait(
+    gex_AD_OpNB_I64(event_AD, count, 
+                    myproc, event_var_ptr,
+                    GEX_OP_GET, 0, 0, 0)
+  );
+}
+
+void caf_event_wait(void *event_var_ptr, int64_t threshold, int segment_boundary, int acquire_fence) {
+  assert(event_AD != GEX_AD_INVALID);
+  assert(event_var_ptr);
+  assert(threshold >= 1);
+
+  if (segment_boundary) {
+    caf_sync_memory();
```

Comment:
During the live code review with @ktras today I realized this line is overkill (issues an acquire fence that duplicates the one below)

Reply:
I've just pushed f7ae50af to adjusted the way this fencing is done, resolving this problem.